pandas/core/dtypes/cast.py

```diff
@@ -319,6 +319,9 @@ def maybe_cast_result_dtype(dtype: DtypeObj, how: str) -> DtypeObj:
         return np.dtype(np.int64)
     elif how in ["add", "cumsum", "sum"] and isinstance(dtype, BooleanDtype):
         return Int64Dtype()
+    elif how == "nunique":
```

Comment:
Do you know where in nunique this is getting cast to date time? Not sure this generic function is the best place to handle this

Reply:
The actual casting is done on the last line of this block:

https://github.com/pandas-dev/pandas/blob/74538104bc04238f69ea6c3a6fbc5468fe02ed12/pandas/core/dtypes/cast.py#L156-L166

This function was added to allow overrides to the default casting behavior based on input datatype and the operation being performed, which I think is what we want to do here.

Reply:
Is that being hit through `_transform_general`? If so seems like something else if off as this should be a reduction function

Reply:
Ah - I think I understand your comment better now: perhaps there is some case in this block that is casting but shouldn't be. In the case at hand, `dtype.kind` is `"M"` and `result.dtype.kind` is `"i"`. Here is an example from the tests where we want to execute the last line:

    df = pd.DataFrame(dict(a=[1], b=pd.to_datetime(["2001"])))
    df.groupby("a")["b"].cummin()

Reply:
Right; ideally we don't have to special case here after we've already cast things. IIUC something seems off with the code path in the first place if we are going through transform machinery to get to this function, as nunique should be a reduction not a transformation

Reply:
The function `maybe_cast_result` is being called by `_transform_fast` for both frames and series:

https://github.com/pandas-dev/pandas/blob/74538104bc04238f69ea6c3a6fbc5468fe02ed12/pandas/core/groupby/generic.py#L1496-L1502

https://github.com/pandas-dev/pandas/blob/74538104bc04238f69ea6c3a6fbc5468fe02ed12/pandas/core/groupby/generic.py#L549-L552

It seems that the typical behavior is to have transform cast: `_transform_should_cast` returns true any time the result is nonempty and the function name is not in `base.cython_cast_blocklist`.

Reply:
Your point is well taken though - if we're taking the result of a reduction and broadcasting to the entire frame/series, why do we need to cast? Taking these casts out, the only test that fails is 

`groupby.transform.test_transform.test_categorical_and_not_categorical_key`.

I'm going to take a deeper look into this case and see what's going on; removing the casting would be a much better solution. Thanks for the comments.